### DISCUSSION TITLE
Why do the docs not recommend the options API?

### DISCUSSION THREAD

**Discussion by bttger** [nuxt/nuxt#18327] [Questions] [answered]
I am wondering why Nuxt 3 discourages to use the Options API. Just because Vue 3 introduced the Composition API, it doesn't mean the Options API got deprecated - quite the opposite, it is still the number one practice and the Vue 3 docs heavily rely on it. 

See https://v3.nuxtjs.org/getting-started/introduction/#comparison

**Comment by kissu** [comment]
Vue3 docs have some work to be done. Evan and his team know it and it is WIP.
Meanwhile, it is recommended to use the Composition API (with `script setup`) for the sole purpose of moving forward.
OptionsAPI is still totally viable indeed but at some point, it is the best to lead people to the more "modern" approach.

Also, I think that more people know how to write things in OAPI than in CAPI, so it's maybe better to show the less known form IMO.

**Reply by bttger** [reply]
I wonder if that is really the general opinion. The Options API has, in my opinion, the huge advantage of clarity. I can get a much faster overview of a component with OAPI. And, of course, the CAPI makes sense in any scenario where I want to share code. It's perfect for that, but I still want to use the OAPI where appropriate. So why explicitly discouraging its use?

**Reply by kissu** [reply]
I don't remember the exact podcast/conf but I do get this info from Evan himself IIRC.

Vue3 is aimed towards a more advanced usage anyways. You don't have to use all features of CAPI but showing examples with only the OAPI is not the best way to go neither IMO. Also, a semi-complex app written in CAPI can be more readable than in OAPI once your components do reach 1000 lines or so.

Having both writings may be cool but it's also a nightmare to maintain I guess. So, rather put the more advanced one.

As of a place to get exactly the opinion of the majority on various topics, [those discussions](https://github.com/vuejs/rfcs/discussions) are probably the best place to look for.

This one in particular: https://github.com/vuejs/rfcs/discussions/378

**Reply by bttger** [reply]
> You don't have to use all features of CAPI but showing examples with only the OAPI is not the best way to go neither IMO.

Don't get me wrong, I'm not against the CAPI 😅 I'm just questioning why Nuxt is explicitly discouraging the use of the OAPI. When I first saw the "not recommended" statement, I was wondering whether there are incompatible features in Nuxt 3. Is there anything incompatible or is it just because Vue is shifting towards CAPI? (not considering the possibly worse TS support)

I agree that Nuxt 3 examples should be written with the CAPI. They must be adapted to each specific use case anyways.

**Reply by kissu** [reply]
We're talking about this one?
![image](https://user-images.githubusercontent.com/5133074/137502086-d4fa20f4-f33c-4786-b556-f29a86af2db6.png)

If yes, I guess that it means "if you want to use only the OAPI, this may not be the most recommended path to take as of right now and you may better use something else".

**Reply by bttger** [reply]
Yes, but does it mean there are features/APIs in Nuxt 3 that are incompatible with the OAPI? I mean the row has a check mark, so I guess not.

**Reply by kissu** [reply]
Some things has been removed, like [the middleware](https://github.com/nuxt/framework/discussions/906#discussioncomment-1476535) and some other things may have some breaking changes (only related to Nuxt specifically AFAIK).

For a more official answer, let's wait somebody from the Core to answer this one!

**Comment by danielroe** [comment, marked as answer]
A few comments:

1. Vue 3 fully supports the Options API, and so of course Nuxt 3 does, out-of-the-box, for all Vue features.

2. This chart is just reflecting the current state of Nuxt 2, Nuxt Bridge, and Nuxt 3 - it's not a roadmap. (So don't be too worried, about this or any other item in the list!)

3. It's true that Nuxt 3 doesn't yet have full Options API support for Nuxt-specific functionality, which is why we don't yet recommend you adopt Nuxt 3 if you're wanting full Options API support for all Nuxt features.

   For example, we do support `head()` in options API, but currently `useAsyncData` is purely usable in `setup()`.

   Sometimes it goes the other way though. For example, `layout` works with Options API but as of yet there's no way to change this in `setup()`. We fully intend to make sure both ways work and have great DX!

4. Apologies that you can't yet try all Nuxt 3 features in the Options API - but don't worry - it's coming soon!

I hope that helps! 😊

**Reply by vedmant** [reply]
@danielroe 2 years later looks like Nuxt 3 still doesn't support options API properly, which just tells about obvious bias towards composition API, which makes life of developers who prefer options API harder.

**Reply by kissu** [reply]
@vedmant mind giving more details as of what doesn't work properly? 🤗

**Reply by vedmant** [reply]
@kissu for example asyncData, rendering error page from it, setting meta from it, etc... Currently the only way to make SSR work is to use hybrid options API / composition API component with setup() function for SSR part.

**Reply by kissu** [reply]
@vedmant I think that sharing a reproduction with the aforementioned issues might help greatly a faster debugging. 🤗

**Reply by vedmant** [reply]
@kissu Issue is created for almost a year now https://github.com/nuxt/nuxt/issues/15616 it's just set as nice-to-have, but this is what basically makes options api not usable with asyncData

**Reply by BananaAcid** [reply]
Some might also be a documentation issue? I am currently trying to figure out, how to use a route-middleware with the options-api ... the docs only have a composition part. @danielroe

**Reply by opalenet-maxime** [reply]
@BananaAcid You can use [definePageMeta](https://nuxt.com/docs/api/utils/define-page-meta) in the setup hook and set [middleware](https://nuxt.com/docs/api/utils/define-page-meta#defining-middleware). 

```
  export default defineNuxtComponent({
    setup() {
      definePageMeta({
        middleware: [],
      });
    }
  })
```

EDIT: Change head to setup
